lib/Parser/Parse.cpp

```diff
@@ -1325,6 +1298,36 @@ ParseNodePtr Parser::CreateProgNodeWithScanner(bool isModuleSource)
     return pnodeProg;
 }
 
+static void InitCallNode(OpCode nop, ParseNodeCall* pnode,
+    ParseNodePtr pnodeTarget, ParseNodePtr pnodeArgs, charcount_t ichMin, charcount_t ichLim)
+{
+    pnode->nop = nop;
+
+    // make Parser::Init public?
```

Comment:
Alternately, make this `Parser::InitCallNode` and have it delegate this first part to `Parser::InitNode`?

Reply:
Yes, could do that too. What is our philosophy on free functions versus members? :)

Reply:
Judging by this file, I think that static class functions would fit in well. I also noticed something similar to what you are doing here:

https://github.com/irinayat-MS/ChakraCore/blob/97412c4447799ed3931f85caabdcaa989564ae33/lib/Parser/Parse.cpp#L713-L718

Perhaps instead of allocating space explicitly in `Create[Super]CallNode`, it would be cleaner to `StaticCreateNodeT<kcbPn[Super]Call>(m_nodeAllocator, ...)` and then `InitCallNode` on the result of that?

Reply:
As we've been generally making a push to move the ParseNodes into a real class hierarchy, I am kind of a fan of adding a method like `ParseNodeCall::Init` which would do this stuff and delegate up to an `ParseNode::Init` for the base class stuff. Or just continue to rely on `InitNode` for that. Could use `pnode->AsParseNodeCall()->Init` or something along those lines. Regardless, I prefer to avoid the duplication of initializing these members in two places.

Reply:
To be clear, this is fine though. Just been thinking about cleanups in this area since Curtis refactored the ParseNodes recently.

Reply:
I find private static methods on a class somewhat weird. Made a step towards class hierarchy with type dependent "Init" methods. Is that what you had in mind?

---
In reply to: [173598566](https://github.com/Microsoft/ChakraCore/pull/4804#discussion_r173598566) [](ancestors = 173598566)

Reply:
Yes, that's what I was thinking about. 👍